src/snark0.ml

```diff
@@ -1066,6 +1066,253 @@ module Make_basic (Backend : Backend_intf.S) = struct
+          proof_system.provide_inputs (Field.Vector.create ()) public_input
+        in
+        let ({Checked.Runner.prover_state= s; _} as state), a =
+          run_proof_system ~run ~input ?system ?eval_constraints ?handler
+            proof_system (Some s)
+        in
+        match s with
+        | Some s -> (s, a, state)
+        | None ->
+            failwith
+              "run_with_input: Expected a value from run_proof_system, got \
+               None."
+
+      let run_unchecked ~run ~public_input ?handler proof_system s =
+        let s, a, _ =
+          run_with_input ~run ~public_input ?handler proof_system s
+        in
+        (s, a)
+
+      let run_checked' ~run ~public_input ?handler proof_system s =
+        let system = R1CS_constraint_system.create () in
+        match
+          run_with_input ~run ~public_input ~system ~eval_constraints:true
+            ?handler proof_system s
+        with
+        | exception e -> Or_error.of_exn e
+        | s, x, state -> Ok (s, x, state)
+
+      let run_checked ~run ~public_input ?handler proof_system s =
+        Or_error.map (run_checked' ~run ~public_input ?handler proof_system s)
+          ~f:(fun (s, x, state) ->
+            let s', x = As_prover.run x (Checked.Runner.get_value state) s in
+            (s', x) )
+
+      let check ~run ~public_input ?handler proof_system s =
+        Or_error.is_ok
+          (run_checked' ~run ~public_input ?handler proof_system s)
+
+      let read_proving_key proof_system =
+        match proof_system.proving_key_path with
+        | Some path -> Some (Proving_key.of_string (In_channel.read_all path))
+        | None -> None
+
+      let write_proving_key proof_system key =
+        match proof_system.proving_key_path with
+        | Some path ->
+            Out_channel.write_all path ~data:(Proving_key.to_string key)
+        | None -> ()
+
+      let read_verification_key proof_system =
+        match proof_system.verification_key_path with
+        | Some path ->
+            Some (Verification_key.of_string (In_channel.read_all path))
+        | None -> None
+
+      let write_verification_key proof_system key =
+        match proof_system.verification_key_path with
+        | Some path ->
+            Out_channel.write_all path ~data:(Verification_key.to_string key)
+        | None -> ()
+
+      let prove ~run ~public_input ?proving_key ?handler proof_system s =
+        let system = R1CS_constraint_system.create () in
+        let _, _, state =
+          run_with_input ~run ~public_input ~system ?handler proof_system s
+        in
+        let {Checked.Runner.input; aux; _} = state in
+        let proving_key =
+          List.find_map_exn
+            ~f:(fun f -> f ())
+            [ (fun () -> proving_key)
+            ; (fun () -> proof_system.proving_key)
+            ; (fun () -> read_proving_key proof_system)
+            ; (fun () ->
+                let {Keypair.pk; vk} = generate_keypair ~run proof_system in
```

Comment:
Should this also call `write_proving_key`?

Reply:
`write_proving_key` is called afterwards, so that a user-provided proving key also gets written. I can pull that out if it's not what we want.